File: scripts/load_test.js
// Load testing script for Thrifts API using k6
// Target: 100 RPS sustained for 30 minutes
// Success criteria: p95 ≤ 1.0s, errors < 0.5%

import http from 'k6/http';
import { check, sleep } from 'k6';
import { Rate, Trend, Counter } from 'k6/metrics';

// Custom metrics
const feedLatency = new Trend('feed_latency');
const eventsLatency = new Trend('events_latency');
const errorRate = new Rate('error_rate');
const feedErrorRate = new Rate('feed_error_rate');
const eventsErrorRate = new Rate('events_error_rate');
const feedsGenerated = new Counter('feeds_generated');
const eventsPosted = new Counter('events_posted');

// Test configuration
export const options = {
  scenarios: {
    // Sustained load: 100 RPS for 30 minutes
    sustained_load: {
      executor: 'constant-rate',
      rate: 100,
      timeUnit: '1s',
      duration: '30m',
      preAllocatedVUs: 50,
      maxVUs: 100,
    },
    
    // Burst test: 200 RPS for 5 minutes  
    burst_test: {
      executor: 'constant-rate',
      rate: 200,
      timeUnit: '1s',
      duration: '5m',
      startTime: '30m',
      preAllocatedVUs: 100,
      maxVUs: 150,
    },
    
    // Events heavy load: 1000 events/s for 10 minutes
    events_heavy: {
      executor: 'constant-rate',
      rate: 1000,
      timeUnit: '1s',
      duration: '10m',
      startTime: '35m',
      preAllocatedVUs: 50,
      maxVUs: 100,
      exec: 'eventsOnly',
    }
  },
  
  thresholds: {
    // SLO targets
    'feed_latency': ['p(95)<1000'], // p95 ≤ 1.0s
    'events_latency': ['p(95)<500'], // Events should be faster
    'error_rate': ['rate<0.005'], // < 0.5% error rate
    'feed_error_rate': ['rate<0.005'],
    'events_error_rate': ['rate<0.005'],
    'http_req_duration': ['p(95)<1000'],
    'http_req_failed': ['rate<0.005'],
  }
};

// Test data generators
const regions = ['Nairobi', 'Mombasa', 'Kisumu', 'Nakuru', 'Eldoret'];
const pages = ['home', 'pdp', 'profile', 'cart', 'checkout'];
const searchTerms = ['dress', 'shoes', 'phone', 'laptop', 'jacket', 'bag', 'watch'];
const eventNames = [
  'page_view', 'product_impression', 'product_click', 'add_to_cart', 
  'search_performed', 'feed_impression'
];

function randomChoice(array) {
  return array[Math.floor(Math.random() * array.length)];
}

function generateUserId() {
  return Math.floor(Math.random() * 10000) + 1;
}

function generateSessionId() {
  return `session_${Math.random().toString(36).substr(2, 9)}`;
}

function generateEventId() {
  return `event_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
}

// Base URL - adjust for your environment
const BASE_URL = __ENV.BASE_URL || 'http://localhost:3000';

export default function() {
  // 70% feeds, 30% events to simulate realistic traffic
  if (Math.random() < 0.7) {
    testFeedGeneration();
  } else {
    testEventIngestion();
  }
  
  // Small random delay to simulate real user behavior
  sleep(Math.random() * 0.5);
}

// Test feed generation endpoint
function testFeedGeneration() {
  const sessionId = generateSessionId();
  const userId = Math.random() > 0.3 ? generateUserId() : null; // 70% authenticated
  const page = randomChoice(pages);
  const region = randomChoice(regions);
  
  // Base feed request
  let payload = {
    page: page,
    session_id: sessionId,
    region: region,
    limit: Math.floor(Math.random() * 20) + 10, // 10-30 items
  };
  
  if (userId) {
    payload.user_id = userId;
  }
  
  // 20% of requests include search
  if (Math.random() < 0.2) {
    if (Math.random() < 0.9) {
      // Text search
      payload.searchType = 'text';
      payload.searchTerm = randomChoice(searchTerms);
    } else {
      // Image search (simulated)
      payload.searchType = 'image';
      payload.imageUrl = 'https://res.cloudinary.com/demo/image/upload/sample.jpg';
    }
  }
  
  // 10% of requests have force_variant for A/B testing
  if (Math.random() < 0.1) {
    payload.force_variant = Math.random() < 0.5 ? 'control' : 'operator';
  }
  
  const startTime = Date.now();
  
  const response = http.post(`${BASE_URL}/api/feed/start`, JSON.stringify(payload), {
    headers: {
      'Content-Type': 'application/json',
    },
  });
  
  const duration = Date.now() - startTime;
  feedLatency.add(duration);
  
  const success = check(response, {
    'feed status is 200': (r) => r.status === 200,
    'feed has feed_id': (r) => {
      try {
        const data = JSON.parse(r.body);
        return data.feed_id && data.feed_id !== 'fallback';
      } catch (e) {
        return false;
      }
    },
    'feed has sections': (r) => {
      try {
        const data = JSON.parse(r.body);
        return Array.isArray(data.sections) && data.sections.length > 0;
      } catch (e) {
        return false;
      }
    },
    'feed has plan_id': (r) => {
      try {
        const data = JSON.parse(r.body);
        return data.plan_id && data.plan_id.length > 0;
      } catch (e) {
        return false;
      }
    },
    'feed response time < 2s': (r) => duration < 2000,
  });
  
  if (!success) {
    feedErrorRate.add(1);
    errorRate.add(1);
  } else {
    feedErrorRate.add(0);
    feedsGenerated.add(1);
    
    // Test pagination if feed was successful
    if (Math.random() < 0.3) { // 30% of feeds test pagination
      testFeedPagination(response);
    }
  }
}

// Test feed pagination
function testFeedPagination(feedResponse) {
  try {
    const feedData = JSON.parse(feedResponse.body);
    
    if (feedData.hasMore && feedData.cursor) {
      const paginationPayload = {
        feed_id: feedData.feed_id,
        cursor: feedData.cursor,
        limit: Math.floor(Math.random() * 15) + 5, // 5-20 items
      };
      
      const startTime = Date.now();
      
      const response = http.post(`${BASE_URL}/api/feed/next`, JSON.stringify(paginationPayload), {
        headers: {
          'Content-Type': 'application/json',
        },
      });
      
      const duration = Date.now() - startTime;
      feedLatency.add(duration);
      
      const success = check(response, {
        'pagination status is 200': (r) => r.status === 200,
        'pagination has products': (r) => {
          try {
            const data = JSON.parse(r.body);
            return data.sections && data.sections[0] && data.sections[0].products.length > 0;
          } catch (e) {
            return false;
          }
        },
      });
      
      if (!success) {
        feedErrorRate.add(1);
        errorRate.add(1);
      } else {
        feedErrorRate.add(0);
      }
    }
  } catch (e) {
    // Ignore pagination errors for malformed feed responses
  }
}

// Test event ingestion endpoint  
function testEventIngestion() {
  const sessionId = generateSessionId();
  const userId = Math.random() > 0.3 ? generateUserId() : null;
  const eventCount = Math.floor(Math.random() * 5) + 1; // 1-5 events per batch
  
  const events = [];
  for (let i = 0; i < eventCount; i++) {
    const event = {
      event_id: generateEventId(),
      session_id: sessionId,
      event_name: randomChoice(eventNames),
      page: randomChoice(pages),
      region: randomChoice(regions),
      timestamp_utc: new Date().toISOString(),
      payload: generateEventPayload(),
    };
    
    if (userId) {
      event.user_id = userId;
    }
    
    events.push(event);
  }
  
  const payload = {
    events: events,
    client_sent_at: new Date().toISOString(),
  };
  
  const startTime = Date.now();
  
  const response = http.post(`${BASE_URL}/api/events`, JSON.stringify(payload), {
    headers: {
      'Content-Type': 'application/json',
    },
  });
  
  const duration = Date.now() - startTime;
  eventsLatency.add(duration);
  
  const success = check(response, {
    'events status is 200': (r) => r.status === 200,
    'events accepted': (r) => {
      try {
        const data = JSON.parse(r.body);
        return data.accepted >= 0 && data.rejected >= 0;
      } catch (e) {
        return false;
      }
    },
    'events response time < 1s': (r) => duration < 1000,
  });
  
  if (!success) {
    eventsErrorRate.add(1);
    errorRate.add(1);
  } else {
    eventsErrorRate.add(0);
    eventsPosted.add(eventCount);
  }
}

// Events-only test function for high-volume scenario
export function eventsOnly() {
  testEventIngestion();
}

// Generate realistic event payload
function generateEventPayload() {
  const payloads = {
    page_view: {},
    product_impression: {
      feed_id: `feed_${Math.random().toString(36).substr(2, 9)}`,
      product_id: Math.floor(Math.random() * 1000) + 1,
      position: Math.floor(Math.random() * 20),
      section: 'grid',
    },
    product_click: {
      feed_id: `feed_${Math.random().toString(36).substr(2, 9)}`,
      product_id: Math.floor(Math.random() * 1000) + 1,
      position: Math.floor(Math.random() * 20),
    },
    add_to_cart: {
      product_id: Math.floor(Math.random() * 1000) + 1,
      quantity: Math.floor(Math.random() * 3) + 1,
      price_cents: Math.floor(Math.random() * 10000) + 1000,
    },
    search_performed: {
      search_term: randomChoice(searchTerms),
      search_type: 'text',
    },
    feed_impression: {
      feed_id: `feed_${Math.random().toString(36).substr(2, 9)}`,
      products: [
        Math.floor(Math.random() * 1000) + 1,
        Math.floor(Math.random() * 1000) + 1,
        Math.floor(Math.random() * 1000) + 1,
      ],
    },
  };
  
  return payloads[randomChoice(eventNames)] || {};
}

// Test teardown
export function teardown(data) {
  console.log('Load test completed');
  console.log(`Feeds generated: ${feedsGenerated.count}`);
  console.log(`Events posted: ${eventsPosted.count}`);
}

File: scripts/load_test.rb
#!/usr/bin/env ruby
# scripts/load_test.rb
# Load testing script for feeds API endpoints

require 'net/http'
require 'json'
require 'benchmark'
require 'concurrent'

class LoadTester
  BASE_URL = 'http://localhost:3000'
  ENDPOINTS = {
    'feeds_start' => '/api/feeds/start',
    'feeds_next' => '/api/feeds/next'
  }

  def initialize(rps:, duration_seconds: 600, error_rate: 0.01)
    @rps = rps
    @duration_seconds = duration_seconds
    @error_rate = error_rate
    @results = []
    @errors = []
    @start_time = Time.now
  end

  def run
    puts "🚀 Starting load test: #{@rps} RPS for #{@duration_seconds} seconds"
    puts "📊 Target endpoints: #{ENDPOINTS.keys.join(', ')}"
    puts "⏰ Started at: #{@start_time}"
    puts "=" * 60

    # Calculate request interval
    interval = 1.0 / @rps
    request_count = @rps * @duration_seconds

    puts "📈 Total requests: #{request_count}"
    puts "⏱️  Request interval: #{interval.round(3)}s"
    puts "=" * 60

    # Start load testing
    start_load_test(interval, request_count)

    # Generate report
    generate_report
  end

  private

  def start_load_test(interval, request_count)
    request_count.times do |i|
      # Simulate error injection
      if rand < @error_rate
        simulate_error_request
      else
        make_request(i)
      end

      # Rate limiting
      sleep(interval) if i < request_count - 1
    end
  end

  def make_request(request_id)
    endpoint = ENDPOINTS['feeds_start']
    payload = generate_test_payload(request_id)

    begin
      response_time = Benchmark.realtime do
        response = make_http_request(endpoint, payload)
        record_result(response, response_time)
      end
    rescue => e
      record_error(e)
    end
  end

  def simulate_error_request
    # Simulate various error conditions
    error_types = [
      -> { raise Net::ReadTimeout.new("Simulated timeout") },
      -> { raise Net::HTTPBadRequest.new("400", "Simulated bad request") },
      -> { raise Net::HTTPServerError.new("500", "Simulated server error") }
    ]

    error_types.sample.call
  rescue => e
    record_error(e)
  end

  def make_http_request(endpoint, payload)
    uri = URI("#{BASE_URL}#{endpoint}")
    http = Net::HTTP.new(uri.host, uri.port)
    http.read_timeout = 10
    http.open_timeout = 5

    request = Net::HTTP::Post.new(uri)
    request['Content-Type'] = 'application/json'
    request.body = payload.to_json

    http.request(request)
  end

  def generate_test_payload(request_id)
    {
      session_id: "load_test_#{request_id}_#{Time.now.to_i}",
      page: ['home', 'pdp', 'profile', 'cart', 'checkout'].sample,
      region: ['Nairobi', 'Mombasa', 'Kisumu'].sample,
      limit: [12, 24, 48].sample,
      pickup_only: [true, false].sample,
      user_id: rand < 0.3 ? rand(1..1000) : nil
    }
  end

  def record_result(response, response_time)
    @results << {
      status: response.code.to_i,
      response_time: response_time,
      timestamp: Time.now,
      success: response.code.to_i < 400
    }
  end

  def record_error(error)
    @errors << {
      error: error.class.name,
      message: error.message,
      timestamp: Time.now
    }
  end

  def generate_report
    puts "\n" + "=" * 60
    puts "📊 LOAD TEST RESULTS"
    puts "=" * 60

    total_requests = @results.size
    successful_requests = @results.count { |r| r[:success] }
    failed_requests = total_requests - successful_requests
    error_requests = @errors.size

    puts "📈 Request Summary:"
    puts "   Total: #{total_requests}"
    puts "   Successful: #{successful_requests}"
    puts "   Failed: #{failed_requests}"
    puts "   Errors: #{error_requests}"
    puts "   Success Rate: #{((successful_requests.to_f / total_requests) * 100).round(2)}%"

    if @results.any?
      response_times = @results.map { |r| r[:response_time] }
      puts "\n⏱️  Response Time Statistics:"
      puts "   Min: #{(response_times.min * 1000).round(2)}ms"
      puts "   Max: #{(response_times.max * 1000).round(2)}ms"
      puts "   Mean: #{(response_times.sum / response_times.size * 1000).round(2)}ms"
      puts "   P50: #{(percentile(response_times, 50) * 1000).round(2)}ms"
      puts "   P95: #{(percentile(response_times, 95) * 1000).round(2)}ms"
      puts "   P99: #{(percentile(response_times, 99) * 1000).round(2)}ms"
    end

    if @errors.any?
      puts "\n❌ Error Summary:"
      error_counts = @errors.group_by { |e| e[:error] }.transform_values(&:size)
      error_counts.each do |error_type, count|
        puts "   #{error_type}: #{count}"
      end
    end

    puts "\n🎯 Performance Targets:"
    puts "   Target RPS: #{@rps}"
    puts "   Target P95: ≤1000ms"
    puts "   Target Error Rate: <0.5%"
    
    # Check if targets are met
    if @results.any?
      p95 = percentile(@results.map { |r| r[:response_time] }, 95)
      error_rate = (error_requests.to_f / total_requests) * 100
      
      puts "\n✅ Target Validation:"
      puts "   P95 ≤1000ms: #{p95 <= 1.0 ? 'PASS' : 'FAIL'} (#{(p95 * 1000).round(2)}ms)"
      puts "   Error Rate <0.5%: #{error_rate < 0.5 ? 'PASS' : 'FAIL'} (#{error_rate.round(3)}%)"
    end

    puts "\n⏰ Test completed at: #{Time.now}"
    puts "⏱️  Total duration: #{Time.now - @start_time} seconds"
  end

  def percentile(array, percentile)
    sorted = array.sort
    index = (percentile / 100.0 * (sorted.length - 1)).round
    sorted[index]
  end
end

# CLI interface
if __FILE__ == $0
  rps = ARGV[0]&.to_i || 100
  duration = ARGV[1]&.to_i || 600

  puts "🔧 Load Test Configuration:"
  puts "   RPS: #{rps}"
  puts "   Duration: #{duration} seconds"
  puts "   Press Ctrl+C to stop early"
  puts "=" * 60

  begin
    tester = LoadTester.new(rps: rps, duration_seconds: duration)
    tester.run
  rescue Interrupt
    puts "\n⏹️  Load test interrupted by user"
    exit 0
  rescue => e
    puts "\n💥 Load test failed: #{e.message}"
    puts e.backtrace.first(5)
    exit 1
  end
end
